docs/proposals/network-policy.md

```diff
@@ -0,0 +1,316 @@
+	// be extended to support other policies in the future.  When set to DefaultDeny,
+	// pods in this namespace are denied ingress traffic by default.  When not defined,
+	// the cluster default ingress policy is applied (currently allow all). 
+	Isolation *IngressIsolationPolicy `json:"isolation,omitempty"` 
+}
+```
+
+```yaml
+kind: Namespace
+apiVersion: v1
+spec:
+  networkPolicy:
+    ingress:
+      isolation: DefaultDeny
+```
+
+The above structures will be represented in v1beta1 as a json encoded annotation like so:
+
+```yaml
+kind: Namespace
+apiVersion: v1
+metadata:
+  annotations:
+    net.beta.kubernetes.io/network-policy: "{\"ingress\": {\"isolation\": \"DefaultDeny\"}}" 
+```
+
+### NetworkPolicy Go Definition
+
+Once a Namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+one or more NetworkPolicy objects should allow any incoming connections that match any
+ingress rule on those NetworkPolicy objects, per the network plugin’s capabilities.
+
+If ingress isolation is not specified on a namespace, then all traffic is allowed to pods in that namespace.
+
+NetworkPolicy objects and the above namespace isolation both act on _connections_ rather than packets.  That is to say that if traffic from pod A to pod B is allowed by the configured
+policy, then the return packets for that connection from B -> A are also allowed, even if policy is in place that would prevent 
+B from initiating a connection to A.  If new network policy is applied that would block an existing connection between two endpoints, the enforcer of policy 
+should terminate and block the existing connection as soon as can be expected by the implementation.
+
+We propose adding a new `/apis/networking` API group for the NetworkPolicy object.
+
+```go
+type NetworkPolicy struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior for this NetworkPolicy.
+  Spec NetworkPolicySpec 
+}
+
+type NetworkPolicySpec struct {
+  // Selects the pods to which this NetworkPolicy object applies.  The array of NetworkPolicyIngressRules below 
+  // is applied to any pods selected by this field. Multiple NetworkPolicy objects can select the 
+  // same set of pods.  In this case, the NetworkPolicyRules for each are combined additively.
+  // This field is NOT optional and follows standard unversioned.LabelSelector semantics.  
+  // An empty PodSelector matches all pods in this namespace.
+  PodSelector unversioned.LabelSelector `json:"podSelector"`
+
+  // List of ingress rules to be applied to the selected pods.
+  // Traffic is allowed to a pod if Namespace.NetworkPolicy.Ingress.Isolation is undefined, 
+  // OR if the traffic source is the pod's local kubelet (for health checks), 
```

Comment:
Yeah, we need to nail down what the right behavior is here.  I think the semantic I'd prefer would be "allow all traffic from pod's local node", but that raises the following:
- Userspace kube-proxy won't support all of this policy out of the box since it obscures source address (this is the case no matter what).
- NodePort services can bypass policy out of the box since they obscure source address.
- pods with `hostNetwork: true` won't work with network policy.

Are we OK with those? I'm not sure.

Alternatively, we could say that _no_ traffic from localhost is allowed and force users to explicitly open ports via NPOs for individual pods on an as needed basis. Or... something else?  I'm not sure the SIG has come to a strong consensus here.